src/crawlee/crawlers/_adaptive_playwright/_adaptive_playwright_crawler.py

```diff
@@ -0,0 +1,425 @@
+from crawlee._utils.docs import docs_group
+from crawlee._utils.wait import wait_for
+from crawlee.crawlers import (
+    AbstractHttpCrawler,
+    AbstractHttpParser,
+    BasicCrawler,
+    BeautifulSoupParserType,
+    ParsedHttpCrawlingContext,
+    PlaywrightCrawler,
+    PlaywrightCrawlingContext,
+    PlaywrightPreNavCrawlingContext,
+)
+from crawlee.crawlers._adaptive_playwright._adaptive_playwright_crawler_statistics import (
+    AdaptivePlaywrightCrawlerStatisticState,
+)
+from crawlee.crawlers._adaptive_playwright._adaptive_playwright_crawling_context import (
+    AdaptivePlaywrightCrawlingContext,
+    AdaptivePlaywrightPreNavCrawlingContext,
+)
+from crawlee.crawlers._adaptive_playwright._rendering_type_predictor import (
+    DefaultRenderingTypePredictor,
+    RenderingType,
+    RenderingTypePredictor,
+)
+from crawlee.crawlers._adaptive_playwright._result_comparator import (
+    create_default_comparator,
+)
+from crawlee.crawlers._beautifulsoup._beautifulsoup_parser import BeautifulSoupParser
+from crawlee.crawlers._parsel._parsel_parser import ParselParser
+from crawlee.statistics import Statistics, StatisticsState
+
+if TYPE_CHECKING:
+    from types import TracebackType
+
+    from typing_extensions import Unpack
+
+    from crawlee.crawlers._abstract_http._abstract_http_crawler import (
+        _HttpCrawlerAdditionalOptions,
+    )
+    from crawlee.crawlers._basic._basic_crawler import _BasicCrawlerOptions
+    from crawlee.crawlers._playwright._playwright_crawler import _PlaywrightCrawlerAdditionalOptions
+
+
+TStaticParseResult = TypeVar('TStaticParseResult')
+TStaticCrawlingContext = TypeVar('TStaticCrawlingContext', bound=ParsedHttpCrawlingContext)
+
+
+class _NonPersistentStatistics(Statistics):
+    """Statistics compliant object that is not supposed to do anything when entering/exiting context.
+
+    To be used in sub crawlers.
+    """
+
+    def __init__(self) -> None:
+        super().__init__(state_model=StatisticsState)
+        self._active = True
+
+    async def __aenter__(self) -> Self:
+        self._active = True
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        exc_traceback: TracebackType | None,
+    ) -> None:
+        self._active = False
+
+
+@docs_group('Classes')
+class AdaptivePlaywrightCrawler(
+    Generic[TStaticCrawlingContext, TStaticParseResult],
+    BasicCrawler[AdaptivePlaywrightCrawlingContext, AdaptivePlaywrightCrawlerStatisticState],
+):
+    """Adaptive crawler that uses both specific implementation of `AbstractHttpCrawler` and `PlaywrightCrawler`.
+
+    It tries to detect whether it is sufficient to crawl without browser (which is faster) or if
+    `PlaywrightCrawler` should be used (in case previous method did not work as expected for specific url.).
+
+    # TODO: Add example
```

Comment:
Forgotton TODO? If it is intented and it is going to happen in the future please link as issue to it.